Add a server side Get operation

smarterclayton opened the pull request:
Implement proposal kubernetes/community#363

```release-note
The Kubernetes API supports retrieving tabular output for API resources via a new mime-type `application/json;as=Table;v=v1alpha1;g=meta.k8s.io`.  The returned object (if the server supports it) will be of type `meta.k8s.io/v1alpha1` with `Table`, and contain column and row information related to the resource.  Each row will contain information about the resource - by default it will be the object metadata, but callers can add the `?includeObject=Object` query parameter and receive the full object.  In the future kubectl will use this to retrieve the results of `kubectl get`.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40848)
<!-- Reviewable:end -->

smarterclayton commented:
@kubernetes/rh-ux if you had opinions about this in a console (this would be the "other resources" default table)

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -158,6 +159,9 @@ type Store struct {
 	ExportStrategy rest.RESTExportStrategy
+	// TableConvertor is an optional interface for transforming items or lists
+	// of items into tabular output. If unset, the default will be used.
+	TableConvertor rest.TableConvertor
```
Barring `ExportStrategy`, the rest of this interface is about managing basic CRUD.  Can you keep the layers separate so we have one layer for CRUD which can be wrapped by something which can do a TableConversion separately?

smarterclayton replied:
ExportStrategy and Tables are the same.  Store isn't about CRUD, it's about REST.  Export/Table are Representations of State.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+	fn := func(obj runtime.Object) error {
+		m, err := meta.Accessor(obj)
+		if err != nil {
+			// TODO: skip objects we don't recognize
```
A silent skip seems wrong.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+		}
+	}
+	table.Headers = []metav1alpha1.TableListHeader{
+		{Name: "Cluster Name", Type: "string", Description: swaggerMetadataDescriptions["clusterName"]},
```
we don't normally show this today, right?

smarterclayton replied:
Only in federation get.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+		{Name: "Cluster Name", Type: "string", Description: swaggerMetadataDescriptions["clusterName"]},
+		{Name: "Namespace", Type: "string", Description: swaggerMetadataDescriptions["namespace"]},
+		{Name: "Name", Type: "string", Description: swaggerMetadataDescriptions["name"]},
+		{Name: "Created At", Type: "date", Description: swaggerMetadataDescriptions["creationTimestamp"]},
```
Most resources don't normally show this today, right?

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+			return nil, err
+		}
+	}
+	table.Headers = []metav1alpha1.TableListHeader{
```
Seems like if you're here, you should be including a column indicating the type.  Similar to how to we do with generic gets now.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+
+var swaggerMetadataDescriptions = metav1.ObjectMeta{}.SwaggerDoc()
+
+func (defaultTableConvertor) ConvertToTableList(ctx genericapirequest.Context, object runtime.Object, tableOptions runtime.Object) (*metav1alpha1.TableList, error) {
```
I seem to be missing where the `tableOptions` are being cast/used.  This is where I'd put things like `--wide`, `--show-kind`, and `--show-labels`, right?

smarterclayton replied:
The default retrieval will return everything for those (wide is modeled as columns that are secondary priority, kind and labels will be returned on each row as partial object metadata).

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/table.go:
```diff
@@ -0,0 +1,105 @@
+	return &table, nil
+}
+
+func trimColumn(column int, table *metav1alpha1.TableList) bool {
```
`trimEmptyColumn`

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/printer/printer.go:
```diff
@@ -0,0 +1,161 @@
+		Wide:      true,
+	}
+	buf := &bytes.Buffer{}
+	args := []reflect.Value{reflect.ValueOf(obj), reflect.ValueOf(buf), reflect.ValueOf(options)}
```
Well this looks weird.

smarterclayton replied:
the generic printer is an affordance to get the right logic in place and then selectively refactor each printer.

smarterclayton replied:
With my changes, I'm going to replace this with a method on the printer that returns a Table.  So you'd get a PrintHandler and call `table := handler.PrintTable(object)`.  And then just return the table.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/printer/printer.go:
```diff
@@ -0,0 +1,161 @@
+		// TODO: this uses more memory than it has to, as we refactor printers we should remove the need
+		// for this.
+		args[0] = reflect.ValueOf(obj)
+		resultValue := handler.printFunc.Call(args)[0]
```
There has to be a better way.

smarterclayton replied:
The correct way is to refactor the printer to return table list - my plan was to get everything cut over to generic, then go back and refactor the printers.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/printer/printer.go:
```diff
@@ -0,0 +1,161 @@
+	return table, nil
+}
+
+func tabbedLineToCells(data []byte, expected int) ([]interface{}, []byte) {
```
test worthy

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/printer/printer.go:
```diff
@@ -0,0 +1,161 @@
+	return nil
+}
+
+func (a *PrinterAdapter) ConvertToTableList(ctx genericapirequest.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1alpha1.TableList, error) {
```
testworthy

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:
```diff
@@ -822,7 +832,8 @@ func UpdateResource(r rest.Updater, scope RequestScope, typer runtime.ObjectType
 		}
-		responsewriters.WriteObject(status, scope.Kind.GroupVersion(), scope.Serializer, result, w, req.Request)
+
+		transformResponseObject(ctx, scope, req, res, status, result)
```
This is actually a `writeObject` call which may optionally transform?  Seems misnamed  and perhaps in the wrong spot.  Why not place in the `responsewriters` package?

smarterclayton replied:
It contains logic that implements the rest protocol.  It calls into storage as needed.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -0,0 +1,75 @@
+				return
+			}
+
+			// renegotiate under the internal version
```
Why?  Seems weird that we'd get an encoder for anything based on an internal version.

bgrant0607 commented:
cc @kubernetes/sig-cli-pr-reviews @kubernetes/sig-api-machinery-pr-reviews

smarterclayton commented:
Updated to include a new field per item `object`.  By default, it is `PartialObjectMetadata` which is an object that only includes `ObjectMeta` (i.e. the thing we want for garbage collection).  If you pass `includeObject=true`, you get the original object.  Clients that want to show labels will always get them.  Clients that need to do JSONPath simply ask for object.

```
$ curl http://localhost:8080/api/v1/services?includeObject=true -H "Accept: application/yaml;as=TableList;v=v1alpha1;g=meta.k8s.io"
apiVersion: meta.k8s.io/v1alpha1
headers:
- description: |-
    Namespace defines the space within each name must be unique. An empty namespace is equivalent to the "default" namespace, but "default" is the canonical representation. Not all objects are required to be scoped to a namespace - the value of this field for those objects will be empty.

    Must be a DNS_LABEL. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/namespaces
  name: Namespace
  type: string
- description: 'Name must be unique within a namespace. Is required when creating
    resources, although some resources may allow a client to request the generation
    of an appropriate name automatically. Name is primarily intended for creation
    idempotence and configuration definition. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/identifiers#names'
  name: Name
  type: string
- description: |-
    CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC.

    Populated by the system. Read-only. Null for lists. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
  name: Created At
  type: date
items:
- cells:
  - default
  - kubernetes
  - 2017-04-15T05:11:58Z
  object:
    apiVersion: v1
    kind: Service
    metadata:
      creationTimestamp: 2017-04-15T05:11:58Z
      labels:
        component: apiserver
        provider: kubernetes
      name: kubernetes
      namespace: default
      resourceVersion: "6"
      selfLink: /api/v1/namespaces/default/services/kubernetes
      uid: 0d0b6df9-219a-11e7-8217-7831c1b76042
    spec:
      clusterIP: 10.0.0.1
      ports:
      - name: https
        port: 443
        protocol: TCP
        targetPort: 443
      sessionAffinity: ClientIP
      type: ClusterIP
    status:
      loadBalancer: {}
kind: TableList
metadata: {}
```

smarterclayton committed:
Add priority and condition to column definitions

Allow the server to indicate additional information along with each row
and table.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -26,20 +26,27 @@ import (
 )
 
+type PrintHandler interface {
+	Handler(columns, columnsWithWide []string, printFunc interface{}) error
```
This is really the best we can offer? `printFunc interface{}`?

smarterclayton replied:
It's because the functions are strongly typed.  I think it's correct enough, we need to strongly validate these in the registration method which I plan on adding eventually.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
+		columns:         columns,
+		columnsWithWide: columnsWithWide,
+	}
+	if err := ValidateRowPrintHandlerFunc(printFuncValue); err == nil {
```
Just going top to bottom, so I haven't found callers, but given an if-block like this I would have expected two different functions instead of one which takes a "sometimes X or sometimes Y" argument.

smarterclayton replied:
I will probably collapse the registration method to one of two paths.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
+// TableHandler adds a print handler with a given set of columns to HumanReadablePrinter instance.
+// See validatePrintHandlerFunc for required method signature.
+func (h *HumanReadablePrinter) TableHandler(columnDefinitions []metav1alpha1.TableColumnDefinition, printFunc interface{}) error {
+	printFuncValue := reflect.ValueOf(printFunc)
```
move closer to usage.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
+	}
+
+	objType := printFuncValue.Type().In(0)
+	h.handlerMap[objType] = entry
```
I would expect an error or panic on overwriting this value to help coders find problems early.

smarterclayton replied:
I'll return an error on double registration.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
 // See validatePrintHandlerFunc for required method signature.
```
comment is now inaccurate.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
+// It must be of the following type:
+//  func printFunc(object ObjectType, options PrintOptions) ([]metav1alpha1.TableRow, error)
+// where ObjectType is the type of the object that will be printed, and the first
+// return value is an array of columns that match the headers.
```
this comment doesn't appear to match the `[]metav1alpha1.TableRow` type.

smarterclayton replied:
Hrm - array of columns = array of cells

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -95,26 +104,90 @@ func (h *HumanReadablePrinter) EnsurePrintHeaders() {
 // See validatePrintHandlerFunc for required method signature.
 func (h *HumanReadablePrinter) Handler(columns, columnsWithWide []string, printFunc interface{}) error {
```
Since we're in here rewriting anyway, why not fix the signature to register to use an instance of the type and the printer can be `func(runtime.Object) (whatever)` and the printers can hard cast.

smarterclayton replied:
I'd prefer to keep the typed functions.  I don't think runtime.Object is better than interface{}.  And I want two different types - one that takes a writer, and one that generates a table.

Also, printers aren't necessarily required to be about runtime.Object.

deads2k commented on pkg/printers/humanreadable.go:
```diff
@@ -195,6 +268,42 @@ func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) er
 		}
+
+		if !handler.rowPrintFunc.IsNil() {
+			args := []reflect.Value{reflect.ValueOf(obj), reflect.ValueOf(h.options)}
```
this reflection stuff is inscrutable.

smarterclayton replied:
I agree.  On the other hand, I am much more concerned with working correctly in the current code path than completely rewriting printers from scratch now.

deads2k commented on pkg/printers/internalversion/printers.go:
```diff
@@ -104,27 +107,37 @@ var (
 
 // AddHandlers adds print handlers for default Kubernetes types dealing with internal versions.
-func AddHandlers(h *printers.HumanReadablePrinter) {
-	h.Handler(podColumns, podWideColumns, printPodList)
```
podcolumns still used for something?

smarterclayton replied:
I'm going to nuke it - I wanted to avoid initializing all the columns as constants anyway.

deads2k commented on pkg/printers/internalversion/printers.go:
```diff
@@ -104,27 +107,37 @@ var (
+		{Name: "Status", Type: "string", Description: "The aggregate status of the containers in this pod."},
+		{Name: "Restarts", Type: "string", Description: "The number of times the containers in this pod have been restarted."},
+		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
+		{Name: "IP", Type: "string", Priority: 1, Description: apiv1.PodStatus{}.SwaggerDoc()["podIP"]},
```
So this priority is logically `wide`.  Is that true for all `priority == 1`?  If so, constant please.

smarterclayton replied:
I kind of think wide is != priority 0.  Open for discussion.  I think long term you'd want to remove columns in descending priority until you fit in the available space (for whatever definition of space is).  I did *not* want to have a `wide bool`.

deads2k commented on pkg/printers/internalversion/printers.go:
```diff
@@ -104,27 +107,37 @@ var (
 	for _, pod := range podList.Items {
-		if err := printPodBase(&pod, w, options); err != nil {
-			return err
+		r, err := printPodBase(&pod, options)
```
Why does this `printPodBase` instead of `printPod`?  Seems like the general case of `FooList` is to call the row printer for `Foo` and aggregate them.

deads2k replied:
> Why does this printPodBase instead of printPod? Seems like the general case of FooList is to call the row printer for Foo and aggregate them.

And if that is case, why are specifying this function at all?

smarterclayton replied:
I think printPodBase never got refactored.  Will fix it here.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go:
```diff
@@ -89,6 +90,35 @@ func Accessor(obj interface{}) (metav1.Object, error) {
+		return &metav1alpha1.PartialObjectMetadata{ObjectMeta: *t}
+	default:
+		return &metav1alpha1.PartialObjectMetadata{
+			ObjectMeta: metav1.ObjectMeta{
```
This looks prone to slipping.  Seems like separate function and some kind of fuzzer or reflective test is in order.

smarterclayton replied:
Yeah, I'm going to flesh this out after I get buy in.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -81,12 +85,12 @@ const (
 // TableOptions are used when a Table is requested by the caller.
 type TableOptions struct {
-	v1.TypeMeta `json:",inline"`
+	v1.TypeMeta `gjson:",inline"`
```
`gjson`?

smarterclayton replied:
Typo, g is my accidental key.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -52,6 +52,10 @@ type TableColumnDefinitions struct {
+	// priority is an integer defining the relative importance of this column compared to others. Lower
+	// numbers are considered higher priority. Columns that may be omitted in limited space scenarios
+	// should be given a higher priority.
+	Priority int32 `json:"priority"`
```
seems weird to have json and not protobuf

smarterclayton replied:
I can't serialize tables to protobuf because of `interface{}` - I don't think we need to do protobuf for 1.7, but I'll do it in the follow up.  I'm not actually sure how to solve cells yet for protobuf.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -0,0 +1,195 @@
+// transformResponseObject takes an object loaded from storage and performs any necessary transformations.
+// Will write the complete response object.
+func transformResponseObject(ctx request.Context, scope RequestScope, req *http.Request, w http.ResponseWriter, statusCode int, result runtime.Object) {
+	// TODO: use returned serializer
```
Why aren't we doing this now?

smarterclayton replied:
I might actually be doing it, will have to go back and look again.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -0,0 +1,195 @@
+		switch {
+
+		case target.Kind == "PartialObjectMetadata" && target.GroupVersion() == metav1alpha1.SchemeGroupVersion:
+			if meta.IsListType(result) {
```
I'm not allowed to get a list of paritalObjectMetadata?  That seems pretty limiting.

smarterclayton replied:
you are, you have to ask for `PartialObjectMetadataList`.  I didn't want to have special assumptions.  `as` is "you get a schema".

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go:
```diff
@@ -43,11 +43,14 @@ import (
 	apitesting "k8s.io/apimachinery/pkg/api/testing"
 	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```
@sttts serialization changes start from here down.  They look ok to me, can you double check them?

smarterclayton committed:
Refactor printers to support rendering as a Table

Return tables from the server.

smarterclayton committed:
Expose a default Table and partial output via Accept headers

All generic registries expose metadata output, and refactor endpoints to
allow negotiation to handle those responses. Add support for
PartialObjectMetadata being returned for objects as well.

smarterclayton commented:
To correctly handling server side filtering, added a new `conditions` field to each row (as mentioned in the proposal).

Refactored the client side code to have a fully compatible implementation of the new `Table` path.  The plan would be to get this in, then complete the rest of the printers, then we can add a client implementation that reads the data.  That could be in 1.8 if necessary, although still possible.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -30,28 +33,33 @@ type Table struct {
 	// See https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#data-types for more.
 	Type string `json:"type"`
-	// format is an optional OpenAPI type definition for this column.
+	// format is an optional OpenAPI type definition for this column. The 'name' format is applied
```
All of these comments should probably start with capital letters...

smarterclayton replied:
No, public godoc is supposed to be in the JSON form.  Then kubectl explain matches the output that a human sees on the outside.

smarterclayton replied:
surprisingly documented here: https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#edit-typesgo

lavalamp replied:
Oh, I see :/

lavalamp replied:
We should probably change it so that things like this are quoted, single ticks or back ticks, so that is clear. Doesn't have to be in this PR of course.

smarterclayton commented:
Yes.  I'll open a 1.8 issue to put a verify script in for API conventions
like this.

On Tue, May 30, 2017 at 1:23 PM, Daniel Smith <notifications@github.com>
wrote:

> *@lavalamp* commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go
> <https://github.com/kubernetes/kubernetes/pull/40848#discussion_r119162570>
> :
>
> >  	// name is a human readable name for the column.
>  	Name string `json:"name"`
>  	// type is an OpenAPI type definition for this column.
>  	// See https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#data-types for more.
>  	Type string `json:"type"`
> -	// format is an optional OpenAPI type definition for this column.
> +	// format is an optional OpenAPI type definition for this column. The 'name' format is applied
>
> We should probably change it so that things like this are quoted, single
> ticks or back ticks, so that is clear. Doesn't have to be in this PR of
> course.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40848#discussion_r119162570>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8eW5uqLf0qDRwny3W-yc9TDsz5Rks5r_FCPgaJpZM4L0uzb>
> .
>

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -60,13 +68,54 @@ type TableRow struct {
+const (
+	// RowCompleted means the underlying resource has reached completion and may be given less
+	// visual priority than other resources.
+	RowCompleted RowConditionType = "Completed"
```
"Completed" doesn't sound universal enough. Having trouble thinking of a better name though.

lavalamp replied:
Active?

smarterclayton replied:
I was actually trying to avoid universal, at least in the "if we add a new condition, it's easy enough for a client to decide to add special logic for it".  Having multiple conditions for different resource types could be acceptable.

I was thinking completed because we have other things that can complete, and it's actually not a hard boundary (filtering a pod that just completed is kind of dumb).  So I had been thinking about transition time.

Also, active is a positive condition we would apply to most things, which means we clutter every row with it.  Advantage of calling out terminal is that it is usually the exceptional case.

lavalamp replied:
All good points.

How about "Uninteresting"?

smarterclayton replied:
Ugh - such a value judgement :)

Inactive?  Terminal?  Hrm

lavalamp replied:
Inactive was going to be my other suggestion, but I'm actually OK with the server providing a value-judgement, since the client isn't going to be opinionated.

lavalamp commented:
I think @mbohlool has started on an API linter, that would be a good place to check things like this.

LGTM - do you want to change the RowCompleted name?

smarterclayton commented:
Let me reflect on the name a bit more and try to find another filtering example.

smarterclayton commented:
Since this is alpha we can always change the name based on feedback later.

smarterclayton commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

smarterclayton commented:
@k8s-bot pull-kubernetes-node-e2e test this

smarterclayton commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-verify test this

smarterclayton commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

smarterclayton commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

droot commented:
@smarterclayton I was testing master and discovered the output of kubectl Get command is not aligned. I ran 'git bisect' to locate the commit where alignment got broken and it turned out to be this PR. I'm new to Kubernetes, so didn't follow this code completely, so thought of commenting it here. This is a critical issue and will block 1.7 release. So please take a look and let me know if you need any help with reproducing it.

smarterclayton commented:
Please open an issue against 1.7 and include the details.

smarterclayton commented:
(And assign to me)

droot commented:
@smarterclayton assigned issue #47093 issue to you.
